src/stream/src/executor/backfill.rs

```diff
@@ -448,6 +502,98 @@ where
             }
         }
     }
+
+    /// Schema
+    /// | vnode | pk | `backfill_finished` |
+    ///
+    /// For `current_pos` and `old_pos` are just pk of upstream.
+    /// They should be strictly increasing.
+    async fn persist_state(
+        epoch: EpochPair,
+        table: &mut StateTable<S>,
+        is_finished: bool,
+        dist_key_in_pk: &[usize],
+        current_pos: &Option<OwnedRow>,
+        old_state: &mut Option<OwnedRow>,
+    ) -> StreamExecutorResult<()> {
+        if let Some(current_pos_inner) = current_pos {
+            let current_state = Self::build_new_state(
+                table.vnode_bitmap(),
+                is_finished,
+                current_pos_inner,
+                dist_key_in_pk,
+            );
```

Comment:
Let us persist all vnodes belonging to this backfill executor instead of just persisting one row, because in my mind if we want to scale backfill executors, new backfill executors need a way to know the offset they should start.

Reply:
> new backfill executors need a way to know the offset they should start.
Could you elaborate on this?

Reply:
```
dev=> create table t(v1 int, v2 int primary key);
CREATE_TABLE
dev=> insert into t values (1, 2), (2, 3), (2, 4);
INSERT 0 3
dev=> flush;
FLUSH
dev=> create materialized view m1 as select * from t;
CREATE_MATERIALIZED_VIEW
dev=> select * from __internal_m1_3_chain_1004
dev-> ;
ERROR:  QueryError: Catalog error: table or source not found: __internal_m1_3_chain_1004
dev=> select * from __internal_m1_5_chain_1004;
 vnode | v2 | t_backfill_finished 
-------+----+---------------------
    75 |  4 | t
   151 |  2 | t
   242 |  3 | t
```

This is what the internal table looks like currently.

Reply:
Sorry. I mean each backfill executor should persist multiple rows (those vnodes belonging to it). In total, we should persist 256 rows if our vnodes number is 256.

Reply:
Even though in the same backfill executor all persisted rows have the same offset, scale-out can benefit from those duplicates.

Reply:
Offline discussion with @chenzl25: Scale in/out of backfill not possible with the following scenario:

Executor 0: vnode 0, 1, 2, 3
```
vnode | offset | is_finished |
0         | xxx      | f      |
...
```

Executor 1: vnode 4, 5, 6, 7
```
vnode | offset | is_finished |
4         | yyy      | f      |
...
```

Scale-in (merge executor vnodes)
Scale-out: vnode 0, 4
Executor 4(new): vnode0, vnode 4, <... more vnodes if scale-in>
```
vnode | offset | is_finished |
0         | xxx      | f      |
4         | yyy      | f      |
```

One idea was to take lowest offset, e.g. if xxx < yyy. Then use xxx. But the state will be inconsistent, we may get duplicate records processed xxx < RECORDs <= yyy.

Reply:
It is fine for now. Since we only mark as "finished" otherwise clean up all the partial state. So the scenario where we have partially complete states upon recovery shouldn't happen. It should be totally empty or totally finished.